feat: add health endpoints, graceful shutdown, and fix race conditions

divakarmanoj opened the pull request:
## Summary

Add health/readiness endpoints for Kubernetes, graceful shutdown support, and fix race conditions.

## Breaking Changes

| Change | Migration |
|--------|-----------|
| `Server.Start()` now returns `error` | Add error handling: `if err := server.Start(addr); err != nil { log.Fatal(err) }` |

## Changes

### Health Endpoints
- Add `/health` - returns 200 if all repos healthy, 503 otherwise
- Add `/ready` - returns 200 if at least one repo refreshed successfully
- Add `/status` - returns detailed status of all repositories

### Graceful Shutdown
- Add `Shutdown()` for graceful shutdown with configurable timeout
- Add `StartWithGracefulShutdown()` for production deployments (handles SIGINT/SIGTERM)

### Server Improvements
- Add HTTP server timeouts (Read: 3min, Write: 3min, Idle: 10min)
- Add per-repository health tracking with `RepositoryStatus` struct
- Add `IsHealthy()`, `IsReady()`, `GetRepositoryStatus()` methods
- Use constant-time comparison for auth key (prevent timing attacks)
- Add `WaitGroup` to track refresh goroutines

### Race Condition Fixes
- Fix race condition on `httpServer` field with mutex
- Add proper ticker cleanup with `defer ticker.Stop()`

## Test plan

- [x] All tests pass with `-race` flag
- [x] Tests for health/ready/status endpoints
- [x] Tests for graceful shutdown
- [x] Concurrent HTTP request tests

divakarmanoj committed:
feat: add health endpoints, graceful shutdown, and fix race conditions

BREAKING CHANGE: Server.Start() now returns error instead of calling Fatal()

- Add /health, /ready, /status endpoints for Kubernetes probes
- Add Shutdown() for graceful shutdown with configurable timeout
- Add StartWithGracefulShutdown() for production deployments
- Add HTTP server timeouts (Read: 3min, Write: 3min, Idle: 10min)
- Add per-repository health tracking with RepositoryStatus struct
- Add IsHealthy(), IsReady(), GetRepositoryStatus() methods
- Fix race condition on httpServer field with mutex
- Use constant-time comparison for auth key to prevent timing attacks
- Add WaitGroup to track refresh goroutines

Tests:
- Add tests for health/ready/status endpoints
- Add tests for graceful shutdown
- Add race condition tests for server
- Add concurrent HTTP request tests

Cidyuk commented on server/server.go:
```diff
@@ -26,57 +53,278 @@ func NewServer(ctx context.Context, repository []source.Repository, refreshInter
+
+	// Repository endpoints
 	for _, repo := range s.Repositories {
+		repo := repo // Capture loop variable to avoid closure bug
```
we are on go 1.22 and this was an issue only for go <= 1.21

divakarmanoj replied:
fixed

Cidyuk commented on server/server.go:
```diff
@@ -26,57 +53,278 @@ func NewServer(ctx context.Context, repository []source.Repository, refreshInter
 	handlers := s.CreateHandlers()
 	handler := etag.Handler(handlers, false)
 	if s.AuthKey != "" {
 		handler = Auth(handler, s.AuthKey)
```
we are wraping the whole handler to Auth - which means we will need to have an auth even for healthcheck
need to verify with @sardine-ai/devops if we are adding the key for the healthcheck in k8s

divakarmanoj replied:
fixed, I dont think its possible now.

divakarmanoj committed:
fix: address PR review comments

- Remove loop variable capture (not needed in Go 1.22+)
- Skip auth for health endpoints (/health, /ready, /status) for K8s probes
- Add test for health endpoints bypassing auth

divakarmanoj committed:
fix: require auth for /status endpoint (only /health and /ready bypass)
